packages/datadog-instrumentations/src/mocha/main.js

```diff
@@ -0,0 +1,678 @@
+      })
+    })
+
+    return run.apply(this, arguments)
+  })
+
+  return Runner
+})
+
+// Used both in serial and parallel mode:
+// In serial mode, this hook is used to set the correct async resource to the test.
+// In parallel mode, the same hook is executed in the worker, so this needs to be repeated in
+// mocha/worker.js
+addHook({
+  name: 'mocha',
+  versions: ['>=5.2.0'],
+  file: 'lib/runnable.js'
+}, (Runnable) => {
+  shimmer.wrap(Runnable.prototype, 'run', run => function () {
+    if (!testStartCh.hasSubscribers) {
+      return run.apply(this, arguments)
+    }
+    const isBeforeEach = this.parent._beforeEach.includes(this)
+    const isAfterEach = this.parent._afterEach.includes(this)
+
+    const isTestHook = isBeforeEach || isAfterEach
+
+    // we restore the original user defined function
+    if (this.fn.asyncResource) {
+      const originalFn = originalFns.get(this.fn)
+      this.fn = originalFn
+    }
+
+    if (isTestHook || this.type === 'test') {
+      const test = isTestHook ? this.ctx.currentTest : this
+      const asyncResource = getTestAsyncResource(test)
+
+      if (asyncResource) {
+        // we bind the test fn to the correct async resource
+        const newFn = asyncResource.bind(this.fn)
+
+        // we store the original function, not to lose it
+        originalFns.set(newFn, this.fn)
+        this.fn = newFn
+
+        // Temporarily keep functionality when .asyncResource is removed from node
+        // in https://github.com/nodejs/node/pull/46432
+        if (!this.fn.asyncResource) {
+          this.fn.asyncResource = asyncResource
+        }
+      }
+    }
+
+    return run.apply(this, arguments)
+  })
+  return Runnable
+})
+
+// Only used in parallel mode (--parallel flag is passed)
+// Used to generate suite events and receive test payloads from workers
+addHook({
+  name: 'workerpool',
+  // mocha@8.0.0 added parallel support and uses workerpool for it
+  // The version they use is 6.0.0:
+  // https://github.com/mochajs/mocha/blob/612fa31228c695f16173ac675f40ccdf26b4cfb5/package.json#L75
+  versions: ['>=6.0.0'],
+  file: 'src/WorkerHandler.js'
+}, (workerHandlerPackage) => {
+  shimmer.wrap(workerHandlerPackage.prototype, 'exec', exec => async function (message, [testSuiteAbsolutePath]) {
+    if (!testStartCh.hasSubscribers) {
+      return exec.apply(this, arguments)
+    }
+    this.worker.on('message', function (message) {
+      if (Array.isArray(message)) {
+        const [messageCode, payload] = message
+        if (messageCode === MOCHA_WORKER_TRACE_PAYLOAD_CODE) {
+          testSessionAsyncResource.runInAsyncScope(() => {
+            workerReportTraceCh.publish(payload)
+          })
+        }
+      }
+    })
+
+    const testSuiteAsyncResource = new AsyncResource('bound-anonymous-fn')
+    testSuiteAsyncResource.runInAsyncScope(() => {
+      testSuiteStartCh.publish({
+        testSuiteAbsolutePath
+      })
+    })
+
+    const result = await exec.apply(this, arguments)
```

Comment:
If the promise is rejected, next lines are not going to be executed, is it something expected?